### DISCUSSION TITLE
Add common properties used in components/elements as arguments/parameters.

### DISCUSSION THREAD

**Discussion by ZeroCool940711** [zauberzeug/nicegui#2538] [Ideas / Feature Requests]
Commonly used things like `alignment`, `colors`, `size`, `padding`, `gap`, `icon`, `tooltip`, `on_click`, `on_hover`, etc should be made available as arguments on every component/element. If possible the same name should be used across elements so we can get used to and find them easily with the IDEs autocomplete, docstrings or code suggestions from tools. After a while it will become part of the flow on how you work with the library and you can easily change one element for another if they share the same arguments. 
Here is an example, lets say we have a Column with a label and two buttons, we want one button and the label to be at the top and the other button to be at the bottom/end of the column like for example in a sidebar, we could just do something like:
```python
with ui.column():
    ui.label("Some label", aligment=ui.alignment.center, size=20)
    ui.button("Button 1", aligment=ui.alignment.top_left, tooltip="Button at the top of our column.", padding=0)
    ui.button("Button 2", aligment=ui.alignment.bottom_left, tooltip="Button at the bottom of our column.", padding=0)
```
Easy to read, righ?
That code could be written even without any documentation at hand and you would just get an idea of what it would look like just by reading the code, just with the normal docstrings on each class and the IDE code completion you can get enough information to do that, it is also intuitive. You also notice that the aligment is in  `ui.aligment` and it will contain a `string` with the position. The same could also be done with just `aligment="top_left"` instead of using `ui.aligment`.

Well, back to realisty. Right now, in order to do that we would need something like:
```python
with ui.column():
    ui.label(title).style("font-size: 20px; align-self: center;")
    ui.button("Button 1").style("align-self: top; gap: 0px").tooltip("Button at the top of our column.")
    ui.button("Button 1").style("align-self: bottom-left; gap: 0px").tooltip("Button at the ottom of our column.")
```
That code looks way more complex than it needs to be, it is also hard to read and unless you know HTML and CSS it will make no sense to you. In order to write those 3 lines for the label and the two buttons it required me spend about 10 minutes searching online to find how to change the position of an element with CSS, how to set the size and the gap, even then I am not sure that code will work until I run it. I could as well just go learn CSS and HTML at that point 🤷‍♂️ 

By having many common things as arguments it would make it possible to work without having to switch constantly between the IDE and the browser to search for something for those of us that dont know much about things like html, css, tailwing or quasar but just python so it would increase productivity a lot, since it will be shown on the IDE itself when hovering over some component and if there is docstrings explaining how to use them it will be easier to know how to do something with little or no effort at all since it will just flow in, you could also just end up learning how to do something that you had no idea how to before by accident just because you tested what an argument does or how it behaves with other arguments, that would be hard or impossible with the way things are right now since you need ot know beforehand how to do something with CSS and HTML. Lastly, it will also reduce the learning curve needed for using NiceGUI, I know a few people that wanted to use NiceGUI and came from Flet and Streamlit but didnt like it because it needed too much HTML and CSS knowledge even for basic stuff.

Note: This discussion could be in part related to #117 since the goal is to try to simplify styling and other things but I think this one goes a bit more in details on how it could be done in a simple and more [Pythonic way ](https://docs.python-guide.org/writing/style/#explicit-code)which is using arguments. Another solution could be having a specific class for styles like what Flet has for their [ElevatedButton](https://flet.dev/docs/controls/elevatedbutton/#style).

**Comment by falkoschindler** [comment]
Thanks for the suggestion, @ZeroCool940711!

I think the idea is related to PR https://github.com/zauberzeug/nicegui/pull/364, which we didn't finish and closed eventually. The challenge is to find out which properties are "common". Some of them are already included in the initializer, like `icon` and `on_click` for buttons, `color` for elements where it makes sense, `size` for icons, ... But if we would add them _all_, the docstring for, e.g., `ui.label` wouldn't be readable anymore, as it would resemble an HTML/CSS specification. Take "alignment" as an example: I personally never thought about changing `align-self` of a button, therefore I wouldn't consider it a common property. But from your perspective it should be part of the public API.

We can circumvent this problem a little bit by moving layout and styling out of the initializer, which is what #362 did for Tailwind and #364 tried for non-Tailwind. But it's very hard to do it right, and I fear the time it takes to maintain such APIs.

Anyway, I'd love to further improve NiceGUI's accessibility, and adding more initializer parameters might be very beneficial. I'm just not yet convinced that there's an easy way to identify and maintain "common" properties, as NiceGUI is used very differently and HTML and CSS are constantly evolving.

**Comment by ZeroCool940711** [comment]
> I think the idea is related to PR #364 , which we didn't finish and closed eventually.

Hi @falkoschindler, as I mentioned on my initial message, this is related to #117 so all the PRs about that might also apply here. 
> The challenge is to find out which properties are "common".

I think you are just thinking too much about this, the answer is actually pretty simple, the arguments we end up adding there will eventually become the commonly used properties since it will be easier than using CSS or HTML properties. We just need to have a starting point and improve on that as times goes and more suggestions are made from the community.

> Take "alignment" as an example: I personally never thought about changing align-self of a button, therefore I wouldn't consider it a common property. But from your perspective it should be part of the public API.

I only used that as an example, I don't even know how to align an element on NiceGUI and have been trying to do that since the first day I started using it, so, as a result I just use other elements to push the element I want to the position I want it to be, which is more work than it should be and I'm not the only one doing that since it appears to be a recurring issue with NiceGUI for those of us that don't know CSS and HTML.

Here are some suggestions on some properties that could be made into arguments and how they should work, most of these are taken from Flet since it does a decent job with some of them:

### Visual and Layout Properties:
- `expand`: When a child element is placed into a [Column](https://nicegui.io/documentation/column) or [Row](https://nicegui.io/documentation/row) you can "expand" it to fill the available space. The `expand` property could be a boolean value (True - expand control to fill all available space) or an integer - an "expand factor" specifying how to divide a free space with other expanded child elements in case there is more than one.
- `visible`: Every element has a `visible` property, which is True by default, this allows the element to be rendered on the page. Setting visible to False completely prevents the element (and all its children, if any) from rendering on a page. Hidden elements cannot be focused or selected with a keyboard or mouse and they do not emit any events.
- `enabled/disabled`: Element is visible, if enabled (default) it can be interacted with and it will trigger events but not when it is disabled. 
- `alignment/horizontal_alignment/text_align`: How the element or its children elements should be placed, it depends on the type of element we are adding, elements like buttons or labels will functionally work different from a `Column` or `Row`. Possible values could be `top_left`, `top_center`, `top_right`, `center_left`, `center`, `center_right`, `bottom_left`, `bottom_center` and `bottom_right`.
- `spacing`: Spacing between controls in a `Column` or `Row` or elements that contain other elements inside.
- `bgcolor/color/text_color`:  Color of the element depending on the type of element. A color value could be a hex value in `#ARGB` format (e.g. `#FFCC0000`), `#RGB` format (e.g. `#CC0000`) or a named color from the `ui.colors` module (list of colors for easy access, PR #2537).
- `margin`: Empty space AROUND the element and child elements. Values could be set for all sides of the element: `left`, `top`, `right` and `bottom`.
- `padding`: Empty space INSIDE a container. The child element is placed inside this padding. Values could be set for padding for all sides of the rectangle: `left`, `top`, `right` and `bottom`.
- `width/height`: Width or height of the element. This value could be a `string` or `int`. If an integer is used and `expand=True` then the value will represent a multiplier to the `expand factor` of the element. If a string is used some possible values could be `10px`, `2rem`, `10%` or any other valid CSS/HTML value.
- `thickness`: For elements such as [ui.separator](https://nicegui.io/documentation/separator#separator) and [ui.splitter](https://nicegui.io/documentation/splitter#splitter), this value can be used to specify the thickness of the line on the element or the size of the element itself.
- `tooltip`: Tooltips provide information which help explain the function of a button or other UI element when the mouse is hovered over the element or long-pressed.
- `icon`: An icon to display.
- `size/text_size`: The size of the element or the text inside of it. For elements that contain both, like a `ui.button`, the `size` will be for the button and the `text_size` will be for the text it contains.
- `leading`: An icon or element to display in front of our element.
- `trailing`: An icon or element to display at the end or after our element.
- `style`: Instance of a Style class that we can use to customize the style of our element.

### Event Related Properties:
- `on_click`: Fires when the user clicks the element. 
- `on_hover/on_focus`: Fires when a mouse pointer enters the element.
- `on_enter/on_exit`: Fires when the mouse enter/exit the element. This is similar to `on_hover/on_focus` but could be useful for more control in some cases.
- `on_dismiss/on_close`: For elements such as [ui.dialog](https://nicegui.io/documentation/dialog#dialog). It fires when the element is dismissed or closed without accepting or canceling an action.
- `on_change/on_update`: Fires when the element is changed or updated, an example would be an [ui.input](https://nicegui.io/documentation/input#text_input) when we type in it.
- `on_scroll`: Fires when the user scrolls inside the element.
- `on_resize`: Fires when an element is resized. 

This should be a good starting point, some of them already exist on some elements, we just need to add it to more elements. No matter how many arguments we have on an element, it will still be better and easier to work with it than having to reference an external source like the CSS/HTML documentation every time we want to add or remove something from our element. 

Let me know what you think about this suggestion.

**Comment by rodja** [comment]
> most of these are taken from Flet since it does a decent job

That is a valid assumption. Flet/Flutter has its own layout engine which works very well. In #364 (which was my attempt to solve #117) I tried to come up with something derived from html/css. But I guess we could implement something similar to the layout behaviour of Flet/Flutter on top of html/css. Let's continue this line of thought!

> `expand`

Ah. Good example. This diverges from the original discussion title "Add common properties used in components/elements as arguments" and highlights the main concern from @falkoschindler. Just adding more arguments is not enough. Some things are better implemented in another way. Btw. we already have something like `expanded`: https://nicegui.io/documentation/space

> `visible`

While visibility can be applied to all elements I wonder if it is common enough to include in the argument list. I find the parameter list of Flet quite overwhelming:

![Screenshot 2024-02-13 at 07 33 20](https://github.com/zauberzeug/nicegui/assets/131391/876b4ec2-13a0-489f-9f93-657081511740)

@falkoschindler and me are more in favour of using builder patterns in addition to parameters. I feel this is the main thing we need to figure out together: when to use builder pattern and when arguments.

> `enabled`

Similar to `visible` but only applies to interactive elements.

I could go on with your other suggestions (which I really like, don't get me wrong). But most fall in the category: are they worth putting into the list of arguments?

**Reply by ZeroCool940711** [reply]
I was only putting some examples, we also don't need to add all of them to every single element, we can only have those that are related to the element itself. We could also use a secondary class for stuff like `Styles` or themes, Flet does that with their buttons styles and other things, instead of adding a bunch of parameters to their button control they go and have a class that handle styling specifically and all the parameters related to that go into that class, the main control/element only has a single `style` parameter which takes an instance of that `Style` class as input. 
Another thing, I mentioned that Flet does some good things but it also has some downsides, one of the biggest issues Flet has is that it tries to make you use too much OOP, everything has to be an object that you have to move around, classes are a must and modularizing your code becomes a bit harder overtime, it constantly breaks the flow of your code and forces you to go out of your way, in that sense I like more the way NiceGUI and Streamlit handle things, I feel like with NiceGUI I can do more prototyping than with Flet, moving things around after you made it is also easier with NiceGUI and that's a good thing.

About the number of parameters in Flet, I personally think that the amount of parameters flet has is not a big deal, it might be related to how your IDE is formatting the parameters and showing them to you, on VS Code using Pylance the formatting is not that bad, even tho it requires some scrolling you can easily find all the information you need right there and then, 
![image](https://github.com/zauberzeug/nicegui/assets/5977640/9a529b79-860b-4957-8a8d-0c14aeb64635)

I personally prefer to have the information I need right there in front of me than having to go out of my way to look for how to do something online which I may or may not find since I might not be able to know what I'm looking for in the first place. Another thing of having the parameters is that we could end up finding how to do something we had no idea how to do or just changing things and end up with something better than what we had by accident, so, it would reward the user's curiosity. I still think we need to look more into this tho, let's continue debating this and hope we get more ideas, maybe there is something we could all agree on.

**Comment by Whhhatttt** [comment]
> Commonly used things like `alignment`, `colors`, `size`, `padding`, `gap`, `icon`, `tooltip`, `on_click`, `on_hover`, etc should be made available as arguments on every component/element. If possible the same name should be used across elements so we can get used to and find them easily with the IDEs autocomplete, docstrings or code suggestions from tools. After a while it will become part of the flow on how you work with the library and you can easily change one element for another if they share the same arguments. Here is an example, lets say we have a Column with a label and two buttons, we want one button and the label to be at the top and the other button to be at the bottom/end of the column like for example in a sidebar, we could just do something like:
> 
> ```python
> with ui.column():
>     ui.label("Some label", aligment=ui.alignment.center, size=20)
>     ui.button("Button 1", aligment=ui.alignment.top_left, tooltip="Button at the top of our column.", padding=0)
>     ui.button("Button 2", aligment=ui.alignment.bottom_left, tooltip="Button at the bottom of our column.", padding=0)
> ```
> 
> Easy to read, righ? That code could be written even without any documentation at hand and you would just get an idea of what it would look like just by reading the code, just with the normal docstrings on each class and the IDE code completion you can get enough information to do that, it is also intuitive. You also notice that the aligment is in `ui.aligment` and it will contain a `string` with the position. The same could also be done with just `aligment="top_left"` instead of using `ui.aligment`.
> 
> Well, back to realisty. Right now, in order to do that we would need something like:
> 
> ```python
> with ui.column():
>     ui.label(title).style("font-size: 20px; align-self: center;")
>     ui.button("Button 1").style("align-self: top; gap: 0px").tooltip("Button at the top of our column.")
>     ui.button("Button 1").style("align-self: bottom-left; gap: 0px").tooltip("Button at the ottom of our column.")
> ```
> 
> That code looks way more complex than it needs to be, it is also hard to read and unless you know HTML and CSS it will make no sense to you. In order to write those 3 lines for the label and the two buttons it required me spend about 10 minutes searching online to find how to change the position of an element with CSS, how to set the size and the gap, even then I am not sure that code will work until I run it. I could as well just go learn CSS and HTML at that point 🤷‍♂️
> 
> By having many common things as arguments it would make it possible to work without having to switch constantly between the IDE and the browser to search for something for those of us that dont know much about things like html, css, tailwing or quasar but just python so it would increase productivity a lot, since it will be shown on the IDE itself when hovering over some component and if there is docstrings explaining how to use them it will be easier to know how to do something with little or no effort at all since it will just flow in, you could also just end up learning how to do something that you had no idea how to before by accident just because you tested what an argument does or how it behaves with other arguments, that would be hard or impossible with the way things are right now since you need ot know beforehand how to do something with CSS and HTML. Lastly, it will also reduce the learning curve needed for using NiceGUI, I know a few people that wanted to use NiceGUI and came from Flet and Streamlit but didnt like it because it needed too much HTML and CSS knowledge even for basic stuff.
> 
> Note: This discussion could be in part related to #117 since the goal is to try to simplify styling and other things but I think this one goes a bit more in details on how it could be done in a simple and more [Pythonic way ](https://docs.python-guide.org/writing/style/#explicit-code)which is using arguments. Another solution could be having a specific class for styles like what Flet has for their [ElevatedButton](https://flet.dev/docs/controls/elevatedbutton/#style).

I agree with you. The only problem why I don't try other python libraries for GUI is CSS/HTML. To make some simple, beautiful window, I will have to write too much text, or watch a large number of tutorials on how to write in CSS.

At the moment I am using FLET, it is really more convenient for me to write applications in it, with a minimum of effort and good results. Imagining that I would have to write a lot of text for each button, or for any other element, drives me crazy. CSS can be automated, just like any other function that has to be written manually for some reason.

But again. There is no need to replace anything. But I believe that such functionality should be present by default so that the user can choose exactly how he will design his application. I would happily use NiceGUI then. With such functionality, I would write a program faster than at the moment. + readability will be good, as well as the ability to quickly edit applications. Overall this is a great idea.

Well, as for the number of parameters, you can always write documentation like flet did. There won't be any problems. But it’s true that the documentation needs to be more detailed than that of FLET. Unfortunately, their documentation is too unclear, which is why you have to ask too much :D

**Comment by rodja** [comment]
Thanks everyone for this valuable discussion. We can all agree on the need to further improve NiceGUIs layout capabilities to make it more intuitive and accessible, while also keeping it maintainable and well-documented. So far, our conversation has illuminated two primary positions: the proposition for "All-in-Argument", versus the current design of providing some functionalities through builder-pattern functions.

### The "All-in-Arguments" Approach
This viewpoint advocates for adding all commonly used properties such as alignment, colors, size, padding, tooltip, etc., directly into component arguments. The main advantages cited include improved intuitiveness, reduced need for CSS/HTML knowledge, and a faster learning curve for new users without deep dives into documentation. Example:

```py
ui.button("Click Me", tooltip="some tooltip description")
```

### The Builder-Pattern Approach
Conversely, the existing approach, which incorporates builder patterns was mainly used to reduce the number of arguments. While we have started implementing some things that way, it is by no means complete. Thinks like alignment, size, padding, ... are currently only  exposed via `.props`, `.classes` or `.style` . Which is not a flaw of this approach itself. Providing builder functions like `.alignment` or `.padding` would also be very intuitive, reduce the need for CSS/HTML knowledge, and provide a fast learning curve for new users without the need of deep dives into documentation. Example:

```py
ui.button("Click Me").tooltip("some tooltip description")
```

I specifically picked the tooltip example to make my point: the builder-pattern also does not require any online-documentation or knowledge of CSS/HTML. New users just need to know that there are configuration options which can be explored by pressing `.` after creating the element. But in my opinion there are several key advantages to this compared to "All-in-Arguments":

#### Modularity and Chainability
The builder pattern allows developers to apply configurations in a modular and chainable manner. This not only enhances code readability but also enables a more intuitive step-by-step application of parameters, making it easier to understand and modify the UI components' behavior without overwhelming constructors.

#### Lazy Configuration
The ability to apply configurations after creation offers more flexibility. This is particularly useful for dynamic UIs that adjust based on runtime data or user interactions, as it allows for changes to be made on the fly.

#### Simplified Documentation
By decoupling component instantiation from configuration, the builder pattern makes the codebase easier to maintain. Writing documentation is more straightforward with the builder pattern, as each method can be documented in isolation, providing clear and concise examples for its use.

**Reply by ZeroCool940711** [reply]
I think we should consider supporting both ways of doing things, one for fast prototyping or beginners to get things done fast and simple, and one way for advanced users with more customization and control over things for when we have a project going and want to improve it further and need something really specific.
The builder pattern will probably not need any change since it's the current way of doing things but we could introduce a `Style` class that take the arguments/parameters like alignment, size, text_color, etc, and it just returns a string that is then passed to `element.style()` itself for compatibility with the current system. We can either modify the existing code to accept a `style` argument to add the extra functionally there and avoid the need for `element.style()` and have it internally just call that, here is an example that I actually implemented as a Proof Of Concept and I can share the code for it or work in a PR to implement this in the core if you guys approve of the idea:
```
from nicegui.Style import Style
from nicegui import app, ui, alignment, color

ui.label("NiceGUI").style(
	Style(
		text_align=alignment.center,
		bgcolor=color.GREEN, 
		text_color=color.BLUE
		)
)

ui.run(native=True, dark=True)
```
![image](https://github.com/zauberzeug/nicegui/assets/5977640/1c82c4e8-3a5a-424e-97aa-d82c6bd6d128)

I formatted the code to make it easier to read but everything on the label could be in a single line and as you can see it depends on PR #2537 . In my test tho I had to do a little hack and wrap the `Style` class in a `str()` when using it inside the `label.style()` since it will only accept a string and not a `Style` object but I think we can fix that easily by allowing it to accept an instance `Style` and not just a string. I also had to import `alignment` an `Style` but ideally we would be able to use them with `ui.alignment` and `ui.Style` to reduce clutter on the imports.

**Reply by ZeroCool940711** [reply]
Forgot to share the `alignment` and `Style` code, it's just this
![image](https://github.com/zauberzeug/nicegui/assets/5977640/32f28775-a011-4efe-beb1-ab9f5aac8250)
![image](https://github.com/zauberzeug/nicegui/assets/5977640/b2d1d2fb-a629-4a50-89df-c22362097efd)

As you can see it can be improved but it works as a POC to show that it's possible to do and wouldn't break the existing system in place.

**Reply by ZeroCool940711** [reply]
I made PR #2552 in case you guys want to test it and give suggestions on what to improve.

**Reply by falkoschindler** [reply]
Thanks for the suggestion, @ZeroCool940711!
But excuse a little provoking question: Isn't this very similar to the already existing [`.tailwind()`](https://nicegui.io/documentation/section_styling_appearance#tailwind_css) method? 🙂

**Reply by ZeroCool940711** [reply]
> But excuse a little provoking question: Isn't this very similar to the already existing [`.tailwind()`](https://nicegui.io/documentation/section_styling_appearance#tailwind_css) method? 🙂

@falkoschindler That's a valid question you have there. I think my suggestion of using the `Style` class is a little different from the `.tailwind()` method. The `Style` class focuses more on using arguments/parameters instead of using the builder pattern and it also tries to simplify things even further than the `.tailwind()` method. The `Style` class, like seen in PR #2552, makes it so we can use the existing `.style()` method so if we have code that was already using CSS/HTML to do the styling of an element we can swap it directly which could make the code more readable and easy to work with in some cases, it also provides an easier to discover option since it would be located in `ui.Style`, it's easy to find it even by accident while scrolling through the code suggestions on VSCode or other IDEs. We could go a bit further and just offer a hint of its existence on the `.style()` method docstring by adding something like `You can also use an instance of ui.Style` even tho the IDE will let the user know that we can use either a `str` or `Style` for the `.style()` method and that should also help to point out the existence of the `ui.Style()` class.

 I do think however that we can use the `.taildwind()` method to improve the code in PR #2552 even further and offer more options.

**Comment by rodja** [comment]
> I think we should consider supporting both ways of doing things, one for fast prototyping or beginners to get things done fast and simple, and one way for advanced users with more customization and control over things for when we have a project going and want to improve it further and need something really specific.

That is exactly NiceGUIs philosophy: make it easy to do common things but also allow customizations through open-ended functions like `.props`, `.classes`, `.style`, `ui.html`, `ui.run_javascript`,...

>  I think my suggestion of using the Style class is a little different from the .tailwind() method. The Style class focuses more on using arguments/parameters instead of using the builder pattern and it also tries to simplify things even further than the .tailwind() method.

I would like the main difference to not be the usage (builder vs parameters) but the simplifications. In my opinion we are looking for an nice API to access layout mechanism which resembles the one from Flutter/Flet. I'll elaborate this on that on the PR #2552 shortly.

**Comment by ffiarpg** [comment]
I totally agree with @ZeroCool940711  and if you look at something like PySimpleGUI it does have long docstrings but it makes it so easy to design GUI inside an IDE without looking up anything in documentation or a browser. It seems to me that half of the benefit of pure python web development should be pythonic development. Currently many elements feel like you need to learn and use css and wrap it in python syntax.

I love the work you've done and it's entirely possible my vision of what NiceGUI should be doesn't match the direction developers want to take it in. 

PySimpleGui
![image](https://github.com/zauberzeug/nicegui/assets/12617741/181bd833-d734-493c-8773-6dbb748144e8)

VS

NiceGUI
![image](https://github.com/zauberzeug/nicegui/assets/12617741/9087b34c-d2ee-44bd-8e6e-4ca74c835ad3)
